### PR TITLE
Patch for PR #412 (Visit Arrow Support Decay Types)

### PR DESCRIPTION
**Pull request by rob-isaac**
These changes were supposed to be in PR #412, but somehow didn't make it (I must have either been on a detached head or messed up a rebase)

### CODE REVIEW

File: libsupport/include/katana/ArrowVisitor.h [hunk 1]
```diff
@@ -212,25 +208,13 @@ std::enable_if_t<
         internal::is_visit_arrow_base_t<Arg0>,
         internal::is_visit_arrow_base_t<Args>...>,
     katana::Result<typename std::decay_t<VisitorType>::ReturnType>>
-VisitArrow(VisitorType&& visitor, const Arg0& arg0, const Args&... args) {
+VisitArrow(VisitorType&& visitor, Arg0&& arg0, Args&&... args) {
```
Comment (witchel) [inline libsupport/include/katana/ArrowVisitor.h:211]: Does this mean that `Arg0` and `Args` are now consumed where they were not before?

**Review by witchel — APPROVED** [review]
One question.

File: libsupport/include/katana/ArrowVisitor.h (hunk 1 above)
Comment (rob-isaac) [reply libsupport/include/katana/ArrowVisitor.h:211]: This means that the arguments are perfectly forwarded to the eventual call to visitor::Call<...>(...). There is no longer an enforcement on the arguments to visitor::Call being const. Does that answer your question?

Comment (witchel) [reply libsupport/include/katana/ArrowVisitor.h:211]: To me, the rvalue-reference type tells me I have to `std::move` my parameter, as opposed to a `const T&`, which tells me the parameter is unchanged in the caller.  But I'm pretty naive when it comes to these C++ constructs.

Comment (rob-isaac) [reply libsupport/include/katana/ArrowVisitor.h:211]: I am also pretty naive but my understanding is that perfect forwarding, which occurs when a function param is rval-ref and it is forwarded to another func, essentially passes the lval/rval type over to the next function that is called. You can use std::move but you dont have to - that just indicates that you are passing a reference that can be modified. Passing an l-value reference means that visitor::Call will also recieve an l-value reference

Comment (amberhassaan) [reply libsupport/include/katana/ArrowVisitor.h:211]: @witchel : When `&&` is added to a template parameter purely ,e.g., `T&&` and not `vector<T>&&`, it becomes a "Universal reference" that captures all types of references of `T` and this enables "Perfect forwarding".

Comment (amberhassaan) [reply libsupport/include/katana/ArrowVisitor.h:211]: @rob-isaac : do you know why we need to separate `Arg0` out? Is that we expect at least one parameter to the function? Just wondering.

**Review by amberhassaan — APPROVED** [review]
looks good.

File: libsupport/include/katana/ArrowVisitor.h (hunk 1 above)
Comment (rob-isaac) [reply libsupport/include/katana/ArrowVisitor.h:211]: @amberhassaan yes it is because we need at least one argument

**Pull request merged**